### ISSUE TITLE
JS sync API is _not_ the fastest?

### ISSUE THREAD

**Issue by lydell** [evanw/esbuild#1689]
The docs state (https://esbuild.github.io/api/#js-specific-details):

> First there is the synchronous API. […] It's also optimal performance-wise if all you need to do is run esbuild and then exit.

All I need to do is run esbuild and then exit – perfect! But I know that the sync API uses a `worker_thread` behind the scenes, and I recently learned those are not free to spawn performance wise. So I measured if the sync API is really the fastest!

Here are the three ways of running esbuild I tested:

```js
// esbuild.transformSync
const newCode = esbuild.transformSync(code, {
  minify: true,
  target: "es5",
  format: "iife",
}).code;

// esbuild.transform
const newCode = (await esbuild.transform(code, {
  minify: true,
  target: "es5",
  format: "iife",
})).code;

// childProcess.spawnSync
const newCode = childProcess.spawnSync(
  "esbuild",
  ["--minify", "--target=es5"],
  {
    input: code,
    encoding: "utf8"
  }
).stdout;
```

These are the results:

| variant                |     time | second run, for fun |
| ---------------------- | -------: | ------------------: |
| esbuild.transformSync  | 77.133ms |             11.95ms |
| esbuild.transform      | 29.758ms |            12.852ms |
| childProcess.spawnSync | 24.375ms |            23.155ms |

Benchmarking is hard, but at least in my case it looks like `esbuild.transformSync` is a clear loser performance-wise when it comes to the “run esbuild once, then exit” use case.

It’s a tie for the second call between `esbuild.transformSync` and `esbuild.transform`. That’s not my use case, but it makes it look like performance is _never_ a reason to choose the sync API. (The valid reasons I can think of for the sync API are: 1) being forced to be synchronous in some context and 2) not wanting to add `async`/`await` a bunch in your code).

Just raw spawning esbuild with `childProcess.spawnSync` _might_ be a _tiny_ bit faster, but the `esbuild.transform()` API is much more convenient to use. That experiment also shows how a repeated call would take the same time, while other approaches become faster.

---

If this results in the docs being updated, I’d suggest looking at this sentence while at it:

> This is the most convenient option because single-threaded JavaScript code has the cleanest syntax.

The paragraph above already says “Because JavaScript is single-threaded” (which is true). What do you mean by “single-threaded JavaScript code has the cleanest syntax”?

**Comment by evanw** [comment]
Yes, that's correct. The docs need to be updated. They were accurate until [version 0.12.11](https://github.com/evanw/esbuild/releases/tag/v0.12.11), when the `worker_threads` approach was made the default behavior. I think the `worker_threads` approach is better for the sync API because the difference matters a lot more when you call it many times, so it's trading off slightly slower in the one-call case vs. much slower in the many-call case. I think maybe I meant non-async instead of single-threaded? Not sure why I said that.
